### PR TITLE
Add an alternative to DeleteVersions that leaves no garbage

### PR DESCRIPTION
**Pull request by klim0v**
The beginning of this issue is that the DeleteVersions method leaves behind a lot of dirty information, which leads to an increase in storage used by the storage.

I first added a method to remove versions up to a certain height, and now I have expanded it to allow removal within a given range.

### CODE REVIEW

File: nodedb.go [hunk 1]
```diff
@@ -237,6 +237,53 @@ func (ndb *nodeDB) DeleteVersionsFrom(version int64) error {
...
+	}
+
+	ndb.mtx.Lock()
+	defer ndb.mtx.Unlock()
+
+	// Delete orphans
+	ndb.traverseOrphans(func(key, hash []byte) {
+		var fromVersion, toVersion int64
+		orphanKeyFormat.Scan(key, &toVersion, &fromVersion)
+		if toVersion < version {
```
Comment (klim0v) [inline nodedb.go:272]:
```suggestion
		if fromVersion < version {
```
Is it possible to do this?
It seems to me that this is not logically correct, but the tests pass. Apparently, this is a bottleneck for me, could you explain to me?

File: tree_test.go [hunk 2]
```diff
@@ -1324,14 +1324,107 @@ func TestLoadVersionForOverwriting(t *testing.T) {
...
-	//Now any key value can be written into the tree
+	// The tree version now is 52 which is equal to latest version.
+	// Now any key value can be written into the tree
 	tree.Set([]byte("key any value"), []byte("value any value"))
 	_, _, err = tree.SaveVersion()
 	require.NoError(err, "SaveVersion should not fail.")
 }
 
-//////////////////////////// BENCHMARKS ///////////////////////////////////////
+func TestDeleteVersionsCompare(t *testing.T) {
```
Comment (klim0v) [inline tree_test.go:1334]: For comparison, you can see my TestDeleteVersionsCompare, which I wrote for debugging, and maybe it should be deleted already. Here is its result:
```
=== RUN   TestDeleteVersionsCompare
    tree_test.go:1365: DeleteVersionsFromInterval database.size 1614
    tree_test.go:1395: DeleteVersion database.size 1614
    tree_test.go:1423: DeleteVersions database.size 1618
--- PASS: TestDeleteVersionsCompare (0.04s)
```

**Comment by erikgrinaker** [comment]
Thanks for contributing! I see you're still making changes to this, and that there a linter issues. Let me know when it's ready for review!

**Comment by klim0v** [comment]
> Thanks for contributing! I see you're still making changes to this, and that there are a few linter issues. Let me know when it's ready for review! Also, maybe convert this to a draft until it's ready.

I see linter crashing in CI / CD config, updated to 1.28 and added `nosec` tag to ignore
_Error: Failed to run: Error: requested golangci-lint version 'v1.27' isn't supported: we support only v1.28.3 and later versions, Error: requested golangci-lint version 'v1.27' isn't supported: we support only v1.28.3 and later versions_

I finished making changes, you can proceed to check! @erikgrinaker

**Comment by klim0v** [comment]
Finished!

**Comment by klim0v** [comment]
@erikgrinaker, I added an alternative, can you check both of them? https://github.com/klim0v/iavl/compare/6933e534d28136c59f00896079d5ea4c952e5097...69e21305a688313e99ff0b2f51379186c497645c

**Comment by klim0v** [comment]
I also suggest to mark the DeleteVersions method as deprecated and try to pass sorted versions from it to the DeleteVersionsInterval.

**Comment by klim0v** [comment]
@alexanderbez, how do you like it?

**Comment by erikgrinaker** [comment]
I've backported a bunch of linter fixes from `master`, so you can rebase to avoid all of the linter stuff.

I'd be interested to find out exactly what this garbage you see left behind by `DeleteVersions` actually is. We have a test `TestRandomOperations` that runs randomized operations against a tree and check invariants, including calling `DeleteVersion` on all versions in random order, and in #326 I've added a test to make sure there are no stray database entries left behind at the end. It doesn't leave any garbage.

Since `DeleteVersions` basically just iterates over all versions and calls `DeleteVersion`, this should be equivalent to what this test does, and it doesn't produce any garbage in the database. Can you elaborate?

**Comment by klim0v** [comment]
> Since `DeleteVersions` basically just iterates over all versions and calls `DeleteVersion`, this should be equivalent to what this test does, and it doesn't produce any garbage in the database. Can you elaborate?

Operation DeleteVersion and DeleteVersions differ in that the commit does not occur on every delete. This is the problem.
Pay attention to the method [deleteOrphans](https://github.com/klim0v/iavl/blob/82251d865e4733c7337f2862d459dae8087fb3cc/nodedb.go#L341) `ndb.getPreviousVersion(version)` will return the values of the previous node that may have been removed in the previous iteration, since it takes this data directly from the persistent node store. As a result, the logic of using this field below is broken in the function passed to the traverseOrphansVersion method

**Review by erikgrinaker — CHANGES_REQUESTED** [review]
Ah, good catch! This should really be implemented using ACID database transactions, to avoid deviations between batched changes and the database state, but that's a larger issue.

This looks great, from what I can tell. Some minor requests:

* Let's rename `DeleteVersionsInterval` to `DeleteVersionsRange`, since we use range as a term for this elsewhere in the code.

* It would be great if you could update `DeleteVersions` to use `DeleteVersionsInterval`.

* It would also be great to add random `DeleteVersionsInterval` calls to `TestRandomOperations`, e.g. by first checking whether a version should be deleted, then randomly choosing to delete either a single or multiple versions, and if multiple delete a random range of e.g. 2-5 versions (which may or may not have gaps).

* The `#nosec` comments are unnecessary now and can be removed.

* `RangeProof.ToProto` and `RangeProofFromProto` are erroneously removed.

**Comment by klim0v** [comment]
> RangeProof.ToProto and RangeProofFromProto are erroneously removed.

the latest version that you uploaded does not find import for `iavlproto`

**Comment by erikgrinaker** [comment]
>>RangeProof.ToProto and RangeProofFromProto are erroneously removed.

>the latest version that you uploaded does not find import for iavlproto

Sorry, my bad - cherry-pick gone wrong. I've reverted this change on `0.14.x` now.

**Comment by klim0v** [comment]
@erikgrinaker, could you help me that I found another subtle place. This happens when you delete blocks again. In the case when I previously deleted data in the old way using DeleteVersions. I got some trash in there. And I clean it up with my new _DeleteVersionsRange_ implementation. But when I call _DeleteVersionsRange_ again at the same interval, I see a tenfold increase in execution time, particularly in the _ndb.traverseOrphansVersion_ location. Why do you think this might be happening? Unfortunately, I can't check the _tree.versions [version]_, because it won't allow me to delete the old garbage. What would you suggest to do?

**Comment by klim0v** [comment]
@erikgrinaker 
why don't we use `&util.Range{Start: start, Limit: end}` in [(db *GoLevelDB) Iterator](https://github.com/tendermint/tm-db/blob/bdf7336cc676c0b7bb518db7e9f247d3a87d7759/goleveldb.go#L161)?

**Comment by erikgrinaker** [comment]
>why don't we use &util.Range{Start: start, Limit: end} in (db *GoLevelDB) Iterator?

Right, this is going to give pretty poor performance. I'm not sure why it's done like this - I know we've had to do some weird things with iterators to get the tm-db semantics to map onto the semantics of the underlying database, but I _think_ in this case it should work.

We have fairly good test coverage for iterators, want to just try using `Range` and see if the tests pass?

**Comment by klim0v** [comment]
> We have fairly good test coverage for iterators, want to just try using `Range` and see if the tests pass?

Yes, all tests pass, including I ran tests from iavl, replacing the database in them with leveldb

**Comment by erikgrinaker** [comment]
Ok, great - did it help with the performance though? I suspect using `Range` and calling `Seek()` are basically the same thing, thus it shouldn't have an effect on performance, but would be good to verify.

**Comment by klim0v** [comment]
> Ok, great - did it help with the performance though? I suspect using `Range` and calling `Seek()` are basically the same thing, thus it shouldn't have an effect on performance, but would be good to verify.

Yes, it helped! Now I'm looking for an explanation of why this happened.

**Comment by klim0v** [comment]
I suppose Seek() could be slow due to the need to go through the entire tree to make sure there is no such key.

**Comment by erikgrinaker** [comment]
Great! It may be that `Seek()` actually iterates. Either way, want to submit a PR for tm-db?

**Comment by klim0v** [comment]
> Great! It may be that `Seek()` actually iterates. Either way, want to submit a PR for tm-db?

Yes, but so far I cannot confirm my words with performance tests to describe this problem...

File: tree_random_test.go [hunk 3]
```diff
@@ -148,12 +149,23 @@ func testRandomOperations(t *testing.T, randSeed int64) {
...
 			versions := getMirrorVersions(diskMirrors, memMirrors)
 			if len(versions) > 2 {
-				deleteVersion := int64(versions[r.Intn(len(versions)-1)])
-				t.Logf("Deleting version %v", deleteVersion)
-				err = tree.DeleteVersion(deleteVersion)
-				require.NoError(t, err)
-				delete(diskMirrors, deleteVersion)
-				delete(memMirrors, deleteVersion)
+				i := r.Intn(len(versions) - 1)
+				if i > versions[0] {
```
Comment (erikgrinaker) [inline tree_random_test.go:153]: I don't understand this logic. `i` here is the index of a random element in the array, why are we comparing that with a version number?

What we should do here, is to randomly pick whether to delete a single or multiple versions (by `deleteRangeChange`), and if multiple the randomly pick a small range to delete (e.g. 2-5 versions).

**Review by erikgrinaker — COMMENTED** [review]
This PR includes a bunch of noisy formatting changes, it would be nice to avoid these since they're unrelated to the main change.

Also, I've released new versions of tm-db and updated the 0.15 and 0.14 branches of IAVL.

**Comment by klim0v** [comment]
> This PR includes a bunch of noisy formatting changes, it would be nice to avoid these since they're unrelated to the main change.
> 
> Also, I've released new versions of tm-db and updated the 0.15 and 0.14 branches of IAVL.

Done ✅

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## Unreleased
+
+### Bug Fixes
+
+- [\#324](https://github.com/cosmos/iavl/pull/324) Fix `DeleteVersions` causing a memory leak. 
+  Add `DeleteVersionsRange` to delete in range (@klim0v)
```
Comment (erikgrinaker) [inline CHANGELOG.md:8]:
```suggestion
- [\#324](https://github.com/cosmos/iavl/pull/324) Fix `DeleteVersions` not properly removing
  orphans, and add `DeleteVersionsRange` to delete a range (@klim0v)
```

**Review by erikgrinaker — APPROVED** [review]
Great stuff, thank you for this! Suggested a few minor nitpicks, otherwise LGTM!

**Comment by klim0v** [comment]
I have added your recommendations, thanks!

**Pull request merged**